spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultPollableMessageSource.java

```diff
@@ -0,0 +1,253 @@
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.binder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+import org.aopalliance.intercept.MethodInterceptor;
+import org.aopalliance.intercept.MethodInvocation;
+
+import org.springframework.aop.framework.ProxyFactory;
+import org.springframework.aop.support.NameMatchMethodPointcutAdvisor;
+import org.springframework.context.Lifecycle;
+import org.springframework.core.AttributeAccessor;
+import org.springframework.integration.core.MessageSource;
+import org.springframework.integration.core.MessagingTemplate;
+import org.springframework.integration.support.AckUtils;
+import org.springframework.integration.support.AcknowledgmentCallback;
+import org.springframework.integration.support.DefaultErrorMessageStrategy;
+import org.springframework.integration.support.ErrorMessageStrategy;
+import org.springframework.integration.support.ErrorMessageUtils;
+import org.springframework.integration.support.StaticMessageHeaderAccessor;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.MessageHandlingException;
+import org.springframework.messaging.support.ChannelInterceptor;
+import org.springframework.retry.RecoveryCallback;
+import org.springframework.retry.RetryCallback;
+import org.springframework.retry.RetryContext;
+import org.springframework.retry.RetryListener;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.util.Assert;
+
+/**
+ * The default implementation of a {@link PollableMessageSource}.
+ *
+ * @author Gary Russell
+ * @since 2.0
+ *
+ */
+public class DefaultPollableMessageSource implements PollableMessageSource, Lifecycle, RetryListener {
+
+	protected static final ThreadLocal<AttributeAccessor> attributesHolder = new ThreadLocal<AttributeAccessor>();
+
+	private final List<ChannelInterceptor> interceptors = new ArrayList<>();
+
+	private final MessagingTemplate messagingTemplate = new MessagingTemplate();
+
+	private MessageSource<?> source;
+
+	private RetryTemplate retryTemplate;
+
+	private RecoveryCallback<Object> recoveryCallback;
+
+	private MessageChannel errorChannel;
+
+	private ErrorMessageStrategy errorMessageStrategy = new DefaultErrorMessageStrategy();
+
+	private BiConsumer<AttributeAccessor, Message<?>> attributesProvider;
+
+	private volatile boolean running;
+
+	public void setSource(MessageSource<?> source) {
+		ProxyFactory pf = new ProxyFactory(source);
+		class ReceiveAdvice implements MethodInterceptor {
+
+			private final List<ChannelInterceptor> interceptors = new ArrayList<>();
+
+			@Override
+			public Object invoke(MethodInvocation invocation) throws Throwable {
+				Object result = invocation.proceed();
+				if (result instanceof Message) {
```

Comment:
Thinking simplification of logic. . . Is it actually possible here for the result to be not-Message?

Reply:
Not in our code but we can't control what a user implementation might do. I tend to prefer a test over a possible CCE.